### PR TITLE
[Feature] Implementation of histogram metric

### LINKED ISSUE #5146: [Proposal] Add Histogram Metrics Implementation
Currently, Doris supports three types of metrics, which are Atomic, LockSimple and CoreLocalCounter. All of them are not quite suitable for statistics analysis. 
Histogram metrics is already in used in many DBMS like kudu, rocksdb and so on. It measures the distribution of values in a stream of data and allow you to measure not just easy things like the min, mean, max, and standard deviation of values, but also quantiles like the median or 95th percentile.
As metrics for statistics analysis, the following interface should be supported:
- `add(uint64_t value)`
- `min()`, `max()`, `num()`, `sum()`
- `average()`, `median()`, `percentile(double p)`, `standard_deviation()`

### PR DESCRIPTION
**Pull request by Skysheepwang**
## Proposed changes

#5146
Add histogram metrics into `util/metrics.h`. The data structure of histogram is implemented in `util/histogram.h`, which could also be used in other situations that in need of histogram. Unit testis added as well.

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] Documentation Update (if none of the other choices apply)
- [x] Code refactor (Modify the code structure, format the code, etc...)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have create an issue on (Fix #5146), and have described the bug/feature there in detail
- [x] Compiling and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [] If this change need a document change, I have updated the document
- [x] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/util/metrics.h [hunk 1]
```diff
@@ -159,6 +160,35 @@ class CoreLocalCounter : public Metric {
     CoreLocalValue<T> _value;
 };
 
+class HistogramMetric : public Metric {
+public:
+    HistogramMetric() { clear(); }
```
Comment (acelyc111) [inline be/src/util/metrics.h:165]: No need to clear() in constructor?

File: be/src/util/histogram.h [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+#include <string>
+#include <vector>
+#include <map>
+#include <atomic>
+
+#include "common/logging.h"
+
+namespace doris {
+
+// Histogram data structure implementation:
```
Comment (acelyc111) [inline be/src/util/histogram.h:29]: These code are almost copied from rocksdb, we have to update LICENSE.txt

File: be/src/util/metrics.h (hunk 1 above)
Comment (Skysheepwang) [reply be/src/util/metrics.h]: Yes, the `_stat.clear()` will be called in `HistogramStat` 's constructor. I will remove it.

**Review by acelyc111 — APPROVED** [review]
LGTM

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**